runtime/binding-mqtt-kafka/src/main/java/io/aklivity/zilla/runtime/binding/mqtt/kafka/internal/stream/MqttKafkaPublishFactory.java

```diff
@@ -227,24 +258,59 @@ private final class MqttPublishProxy
         private OctetsFW clientIdOctets;
         private boolean retainAvailable;
         private int publishFlags;
+        private int packetId;
+        private int qos;
 
         private MqttPublishProxy(
             MessageConsumer mqtt,
             long originId,
             long routedId,
             long initialId,
             long resolvedId,
+            long affinity,
+            MqttKafkaBindingConfig binding,
             String16FW kafkaMessagesTopic,
             String16FW kafkaRetainedTopic,
+            int qos,
             List<Function<String, String>> clients)
         {
             this.mqtt = mqtt;
             this.originId = originId;
             this.routedId = routedId;
             this.initialId = initialId;
             this.replyId = supplyReplyId.applyAsLong(initialId);
-            this.messages = new KafkaMessagesProxy(originId, resolvedId, this, kafkaMessagesTopic);
-            this.retained = new KafkaRetainedProxy(originId, resolvedId, this, kafkaRetainedTopic);
+            this.affinity = affinity;
+            if (qos == MqttQoS.EXACTLY_ONCE.value())
+            {
+                final KafkaMessagesProxy messages = binding.sessions.get(affinity).getQos2PublishStream(kafkaMessagesTopic);
+                if (messages != null)
+                {
+                    messages.delegate = this;
+                    this.messages = messages;
+                }
+                else
+                {
+                    this.messages = new KafkaMessagesProxy(originId, resolvedId, routedId, affinity, this, kafkaMessagesTopic);
+                }
+
+                final KafkaRetainedProxy retained = binding.sessions.get(affinity).getQos2PublishRetainedStream();
+                if (retained != null)
+                {
+                    retained.delegate = this;
+                    this.retained = retained;
+                }
+                else
+                {
+                    this.retained = new KafkaRetainedProxy(originId, resolvedId, routedId, affinity, this, kafkaRetainedTopic);
+                }
+            }
+            else
+            {
+                this.messages =
+                    new KafkaMessagesProxy(originId, resolvedId, routedId, affinity, this, kafkaMessagesTopic);
+                this.retained =
+                    new KafkaRetainedProxy(originId, resolvedId, routedId, affinity, this, kafkaRetainedTopic);
+            }
```

Comment:
I'd like to better understand why this complexity is needed.
Seems like we just want to defer the offset commit to session, everything else stays the same, no?